File: Data Structures & Algorithms/count-number-of-islands/submission-1.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    res += 1
                    stack = []
                    stack.append([i, j])
                    while stack:
                        cell = stack.pop()
                        r = cell[0]
                        c = cell[1]
                        grid[r][c] = '0'
                        if r + 1 < len(grid) and grid[r + 1][c] == '1':
                            stack.append([r + 1, c])
                        if r - 1 >= 0 and grid[r - 1][c] == '1':
                            stack.append([r - 1, c])
                        if c + 1 < len(grid[0]) and grid[r][c + 1] == '1':
                            stack.append([r, c + 1])
                        if c - 1 >= 0 and grid[r][c - 1] == '1':
                            stack.append([r, c - 1])
        return res
        

File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int max_size = 0;
        int rows = grid.size();
        int columns = grid[0].size();

        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < columns; ++j) {
                if (grid[i][j] == 1) {
                    grid[i][j] = 0;
                    int size = 0;
                    stack<pair<int, int>> stack;
                    stack.push({i,j});
                    while(!stack.empty()) {
                        pair<int, int> cell = stack.top();
                        stack.pop();
                        ++size;
                        int r = cell.first;
                        int c = cell.second;

                        if (r + 1 < rows && grid[r + 1][c] == 1) {
                            grid[r + 1][c] = 0;
                            stack.push({r + 1, c});
                        }
                        if (c + 1 < columns && grid[r][c + 1] == 1) {
                            grid[r][c + 1] = 0;
                            stack.push({r, c + 1});
                        }
                        if (r - 1 >= 0 && grid[r - 1][c] == 1) {
                            grid[r - 1][c] = 0;
                            stack.push({r - 1, c});
                        }
                        if (c - 1 >= 0 && grid[r][c - 1] == 1) {
                            grid[r][c - 1] = 0;
                            stack.push({r, c - 1});
                        } 

                    }

                    max_size = max(size, max_size);
                }
            }
        }

        return max_size;
    }
};
